### PR TITLE
Memory graphs real usage

### PR DESCRIPTION
**Pull request by Gundersanne**

### CODE REVIEW

**Comment by garrett** [comment]
I checked this on my system and the stats looked a lot better, but the values on the left axis were off. I'll try to get the binary part compiled again to re-test to be sure.

**Comment by garrett** [comment]
Ideally, the y-axis would be fixed. As this is a bugfix for the memory being displayed, I suppose fixing the labeling could be in another PR?

**Comment by Gundersanne** [comment]
@garrett This was sort of a test of how far I could get to the mockups with jquery flotr. I can spend some more time on the y-axis. I think it's better to fix it in 1 pr since it's all related.

**Comment by garrett** [comment]
As mentioned on IRC, I enabled swap on my system with a swapfile. 

The small overview has a red line:

memory-graph-small

The detailed graph does not show any value:

memory-graph-large

Here's what `free` says:

```
              total        used        free      shared  buff/cache   available
Mem:       20405236     9289480      538544     1358564    10577212     9533244
Swap:        524284        1792      522492
```

...so my system with a ton of RAM is actually using swap a little tiny bit.

---

I agree with your IRC suggestion of turning off the line, at least for the swap. :+1:

**Comment by Gundersanne** [comment]
@garrett without lines:

image

**Comment by garrett** [comment]
Rebased on #10282

**Comment by garrett** [comment]
Here are colors I chose for CPU, weighted on color contrast, lightness contrast, and colorblindness-friendliness:

screenshot_2018-10-18 system - sunny

...it's a lot of purple. At least on my machine. But I specifically avoided blue, as it usually means "link", and I tried to align red == bad and green == good. User and IO are both darkish and then I have kernel and nice as lightish, in a greyscale zebra-ish pattern when desaturated (100% colorblindness).

I also adjusted the order of the legend to be in the same order as the data.

I don't really have much IO weight on my machine, and I have no nice processes at the time of the screenshot. The busy left half is with BlueJeans running from our standup and the right part is my system breathing a sigh of relief from stopping BlueJeans. :wink:

(Memory is _probably_ similar, as I patterned this after the colors and technique I used for the memory graph. However, I think we'll need to adjust it in the other PR as values changed.)

**Comment by garrett** [comment]
Graphs in the networking page are too light for me:

screen shot 2018-10-18 at 16 40 42-fullpage

Note a tiny blip of normal color in the receiving graph.

**Comment by garrett** [comment]
Here's what the too-light graph looks like without the changes in this PR:

screen shot 2018-10-18 at 16 58 55-fullpage

So the problem is unrelated to the changes in this PR... and this is why I had a big problem trying to debug it.

I think the issue stems from the fact that my currently active main interface is the 5th on the interfaces list. How do the colors get chosen for the various interfaces?

File: pkg/lib/plot.js [hunk 1]
```diff
@@ -397,6 +397,131 @@ plotter.plot = function plot(element, x_range_seconds, x_stop_seconds) {
         return self;
     }
 
+    function add_metrics_difference_series(desc, opts) {
```
Comment (mvollmer) [inline pkg/lib/plot.js]: Is this just `add_metrics_difference_series` with `-` instead of `plus`?

File: src/bridge/cockpitmemorysamples.c [hunk 2]
```diff
@@ -66,10 +67,14 @@ cockpit_memory_samples (CockpitSamples *samples)
...
       else if (g_str_has_prefix (line, "Cached:"))
         g_warn_if_fail (sscanf (line + sizeof ("Cached:") - 1, "%" G_GUINT64_FORMAT, &cached_kb) == 1);
+      else if (g_str_has_prefix (line, "MemAvailable:"))
+        g_warn_if_fail (sscanf (line + sizeof ("MemAvailable:") - 1, "%" G_GUINT64_FORMAT, &available_kb) == 1);
+
     }
 
   cockpit_samples_sample (samples, "memory.free", NULL, free_kb * 1024);
   cockpit_samples_sample (samples, "memory.used", NULL, (total_kb - free_kb) * 1024);
+  cockpit_samples_sample (samples, "memory.unavailable", NULL, (total_kb - available_kb) * 1024);
```
Comment (mvollmer) [inline src/bridge/cockpitmemorysamples.c]: I think it would be fine just to fix the way we compute `memory.used`.  No need to introduce a new metric, IMO.

**Comment by mvollmer** [comment]
I think we need to explain why MemAvailable is better than what we used to do.

**Comment by garrett** [comment]
Oh, I typed up this comment yesterday, but GitHub was being weird all day... so here it is, with a few edits:

---

> I think we need to explain why MemAvailable is better than what we used to do.

Agreed. This explanation should also go in the release notes / blog post. It could probably have text similar to this:

## Memory graph improvements

Memory on Linux is often hard to understand and usually presented incorrectly in most tools. There are so many different terms, most of which do not matter to anyone who's not a kernel developer. 

In addition, Cockpit's old memory graphs were unfortunately calculated incorrectly as well.

The new presentation of memory in Cockpit is aligned with the way people think about how RAM is used. It now shows:

1. the amount of RAM actively used by software
2. the amount of RAM used for cache
3. unused RAM
4. swap

On Cockpit's overview, the memory graph displays only the value of active RAM and has a separate, small area to indicate swap usage. The detailed memory view also shows cache.

Why do we not show cache on the overview and de-emphaize it on the detailed memory view? Linux stores often-used files in memory as cache to speed up performance. As cache is quickly reclaimed when needed, it's effectively available RAM. That's why we consider it free space, as Linux does too.

Additional information on can be found on the website [Linux Ate My RAM](https://www.linuxatemyram.com/).

**Comment by martinpitt** [comment]
@mvollmer: As you previously reviewed this, can you please do the final review? This seems to work now.

@Gundersanne: Is the screenshot above still current? Or, I suppose it should rather be attached to #10282 as that changes the design again?

**Comment by garrett** [comment]
@martinpitt Yeah, we need both this PR and #10282 to land before taking screenshots for the release notes. 

This PR is purely for functionality for the memory graphs, and the other applies to all the charts (CPU and memory) and modifies the layout a bit (prevent legend from overlapping the chart), reorder the legend based on color order in the chart, align colors to the PatternFly palette, and be more colorblind-friendly.

File: test/verify/check-metrics [hunk 3]
```diff
@@ -115,7 +115,7 @@ class TestMetrics(MachineCase):
         # check whether we see that in the plot.
         #
         meminfo = read_mem_info(m)
-        mem_used = meminfo['MemTotal'] - meminfo['MemFree']
+        mem_used = meminfo['MemTotal'] - meminfo['MemAvailable']
```
Comment (mvollmer) [inline test/verify/check-metrics]: (Should this be in the previous commit?)

Comment (Gundersanne) [reply test/verify/check-metrics]: oh yea, good spot.

**Comment by mvollmer** [comment]
"lib: Untangle different metrics from plot" was a bit confusing. The commit is about rewriting the different series kinds as ES6 classes. A metric is something like "mem.used".

**Comment by mvollmer** [comment]
I have "Cached+Used" reach into the "Swap" area.  Is that expected?

mem

**Comment by mvollmer** [comment]
> I have "Cached+Used" reach into the "Swap" area. Is that expected?

That was with the old cockpit-bridge, sorry.

**Pull request merged**